csrc/scheduler/normalization_inner_outer.cpp

```diff
@@ -199,31 +459,129 @@ std::pair<int64_t, int64_t> getPersistentBufferSize(
+      fusion,
+      reduction_tvs,
+      InnerOuterPersistentKernelScheduler::threads_per_block_max);
+  int64_t available_smem =
+      (int64_t)dev_prop->sharedMemPerMultiprocessor - smem_overhead;
+  int64_t available_regs = scheduler_utils::register_file_size_full;
+  buffer_params.smem_overhead = smem_overhead;
+
+  // (1) init the buffer_params by putting all the persistent tensors in
+  // registers
+  buffer_params.regs_buffer_size = total_buffer_size;
+  buffer_params.smem_buffer_size = 0;
+
+  // (2) If the available register is larger than current register buffer size,
+  // no need to move buffers to shared memory, return early.
+  if (buffer_params.regs_buffer_size <= available_regs) {
+    buffer_params.has_enough_regs_and_smem = true;
+    return buffer_params;
   }
-  persistent_buffer_size +=
-      normalization_scheduler_utils::partialReductionBufferSize(
-          outer_reduction_tvs, runtime_info);
 
-  int64_t available_persistent_buffer_size =
-      scheduler_utils::register_file_size_full;
+  // (3) Relocate buffers to shared memory until the buffer size in registers is
+  // within the allowable limit.
+
+  // (3.1) Sort the candidate persistent buffers, based on its read/write usage.
+  // Put the buffer that is less frequently used in the front.
+  const auto sorted_candidate_tvs = buffer_params.project_to_input
+      ? sortProjectableBufferInputs(
+            fusion,
+            persistent_buffer_info.persistent_buffers,
+            persistent_buffer_info.projectable_buffer_inputs,
+            outer_broadcast_tvs,
+            reduction_tvs,
+            runtime_info)
+      : sortPersistentBuffers(persistent_buffer_info.persistent_buffers);
+
+  // (3.2) Before this loop, all buffers are in registers.
+  // Try to move buffer from register to shared memroy. The visiting order is
+  // from the first buffer to the last buffer in the sorted_candidate_tvs.
+  // After one buffer is moved to shared memory, the buffer size in registers
+  // and shared memory are updated accordingly. Break if required register and
+  // shared memory are lower than limit or shared memory exceeds the limit.
+  int64_t n_smem_buffer = -1;
+  const int n_buffers = (int)sorted_candidate_tvs.size();
+  for (int i = 0; i < n_buffers; i++) {
+    // start from the first buffer, which has the smallest read/write latency
+    // move it to shared memory has a smaller impact on performance than moving
+    // other more frequently accessed buffers.
+    auto current_tv = sorted_candidate_tvs[i];
+
+    // calculate the size of this buffer & reduce the register buffer size
+    int64_t tv_buffer_size_regs =
+        scheduler_utils::getPersistentBufferSizeOfTensor(
+            current_tv, runtime_info, persistent_buffer_info);
+    buffer_params.regs_buffer_size -= tv_buffer_size_regs;
+
+    // round up the buffer size to shared memory & increase the shared memory
+    // buffer size
+    int64_t tv_buffer_size_smem = roundUpSharedMemory(
```

Comment:
Actually, I guess the register usage is underestimated as it doesn't seem to account for the redundant nature of broadcast allocation.

Reply:
Talked with Liqiang offline. This was my misunderstanding. Since the tensor itself doesn't have broadcast domains, there's no concern of underestimation.